docs/proposals/11-cluster-health.md

```diff
@@ -0,0 +1,205 @@
+  It is probably because the `kube-apiserver` endpoint is not reachable externally due to loadbalancer or some other network issue.
+  - The `root` [command](https://github.com/gardener/dependency-watchdog/blob/master/cmd/root.go) of the `dependency-watchdog` watches the Kubernetes `endpoints` of `etcd` and `kube-apiserver` of all the shoots (hosted in the seed cluster).
+  If any of the endpoints transition from unhealthy to healthy it restarts (deletes) the corresponding [downstream](https://github.com/gardener/gardener/blob/master/charts/seed-bootstrap/charts/dependency-watchdog/templates/endpoint-configmap.yaml#L41) [dependent](https://github.com/gardener/gardener/blob/master/charts/seed-bootstrap/charts/dependency-watchdog/templates/endpoint-configmap.yaml#L16) `pods` but only if they are in `CrashLoopBackoff`.
+  This is to speed up the cluster recovery from `etcd` or `kube-apiserver` unavailability.
+  - The `etcd` `readinessProbe` [points](https://github.com/gardener/gardener/blob/master/charts/seed-controlplane/charts/etcd/templates/etcd-statefulset.yaml#L43) to the health of the [etcd-backup-restore](https://github.com/gardener/etcd-backup-restore) [sidecar](https://github.com/gardener/gardener-extensions/blob/master/pkg/webhook/controlplane/etcd.go#L44).
+  This makes the `etcd` of the control-plane of the shoot clusters unavailable if the backup infrastructure is unavailable for any reason (even if the `etcd` container itself is healthy).
+  - The `machine-controller-manager` can in principle (it currently doesn't) notice issues in the cloud-provider that other Kubernetes and or Gardener controllers do not notice such as machines not joining in the given timeout period, cloud-provider API requests failing etc.
+
+The distributed nature of such information and action is in itself not a *bad thing*.
+But if the information is implicit and isolated within different controllers then there is a hard coupling between the sources of the information and the corrective actions that are taken which **is** a *bad thing*.
+
+## Goals
+
+- As a distributed system, the components that monitor health conditions that have cluster-wide implication should be decoupled from the components that take remedial action.
+- As a distributed system, it should be possible to dynamically deploy/undeploy new components that monitor some (possibly new) health conditions that have cluster-wide implications with minimal impact.
+The information to determine the such health conditions could be sourced from within the cluster, outside the cluster or some combination.
+- As a distributed system, it should be possible to dynamically deploy/undeploy new components that remedy some health conditions that have cluster-wide implications with minimal impact.
+- As a controller that checks for some health conditions that have cluster-wide implications, it should be possible to post the health condition status in a standard way.
```

Comment:
> Why post? It might be pull based, or both?

This is not about Prometheus, Dynatrace of any such metrics posting. I am not proposing any new monitoring system. 
This goal is just about a controller that monitors some health conditions of the cluster and updates the health condition in some standard way/location such as `status.conditions` in some resource which can be watched by other controllers which can take their own domain-specific action.